server/src/main/java/org/opensearch/action/search/SearchQueryCategorizingVisitor.java

```diff
@@ -39,29 +60,46 @@ public SearchQueryCategorizingVisitor(SearchQueryCounters searchQueryCounters) {
+        handlers.put(DistanceFeatureQueryBuilder.class, searchQueryCounters.distanceFeatureCounter);
+        handlers.put(ExistsQueryBuilder.class, searchQueryCounters.existsCounter);
+        handlers.put(FieldMaskingSpanQueryBuilder.class, searchQueryCounters.fieldMaskingSpanCounter);
+        handlers.put(FuzzyQueryBuilder.class, searchQueryCounters.fuzzyCounter);
+        handlers.put(GeoBoundingBoxQueryBuilder.class, searchQueryCounters.geoBoundingBoxCounter);
+        handlers.put(GeoDistanceQueryBuilder.class, searchQueryCounters.geoDistanceCounter);
+        handlers.put(GeoPolygonQueryBuilder.class, searchQueryCounters.geoPolygonCounter);
+        handlers.put(GeoShapeQueryBuilder.class, searchQueryCounters.geoShapeCounter);
+        handlers.put(IntervalQueryBuilder.class, searchQueryCounters.intervalCounter);
+        handlers.put(MatchAllQueryBuilder.class, searchQueryCounters.matchallCounter);
+        handlers.put(PrefixQueryBuilder.class, searchQueryCounters.prefixCounter);
+        handlers.put(ScriptQueryBuilder.class, searchQueryCounters.scriptCounter);
+        handlers.put(SimpleQueryStringBuilder.class, searchQueryCounters.simpleQueryStringCounter);
+
+        return handlers;
     }
 
     public void accept(QueryBuilder qb) {
-        if (qb instanceof BoolQueryBuilder) {
-            searchQueryCounters.boolCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof FunctionScoreQueryBuilder) {
-            searchQueryCounters.functionScoreCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof MatchQueryBuilder) {
-            searchQueryCounters.matchCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof MatchPhraseQueryBuilder) {
-            searchQueryCounters.matchPhrasePrefixCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof MultiMatchQueryBuilder) {
-            searchQueryCounters.multiMatchCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof QueryStringQueryBuilder) {
-            searchQueryCounters.queryStringQueryCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof RangeQueryBuilder) {
-            searchQueryCounters.rangeCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof RegexpQueryBuilder) {
-            searchQueryCounters.regexCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof TermQueryBuilder) {
-            searchQueryCounters.termCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
-        } else if (qb instanceof WildcardQueryBuilder) {
-            searchQueryCounters.wildcardCounter.add(1, Tags.create().addTag(LEVEL_TAG, level));
+        Counter counter = queryHandlers.get(qb.getClass());
+        if (counter != null) {
+            counter.add(1, Tags.create().addTag(LEVEL_TAG, level));
```

Comment:
Nice refactoring! Although, it will be even to encapsulate this information within `SearchQueryCounters`. The visitor can just pass the query builder class and `SearchQueryCounters` can make the decision to increment the corresponding counter. That will also ensure that these counters are not public